fedimint-api-client/src/api/mod.rs

```diff
@@ -1032,6 +1062,575 @@ where
+    fn all_peers(&self) -> &BTreeSet<PeerId> {
+        &self.peers
+    }
+
+    fn self_peer(&self) -> Option<PeerId> {
+        self.admin_id
+    }
+
+    fn with_module(&self, id: ModuleInstanceId) -> DynModuleApi {
+        ReconnectFederationApi {
+            peers: self.peers.clone(),
+            admin_id: self.admin_id,
+            module_id: Some(id),
+            connections: self.connections.clone(),
+        }
+        .into()
+    }
+
+    async fn request_raw(
+        &self,
+        peer_id: PeerId,
+        method: &str,
+        params: &ApiRequestErased,
+    ) -> JsonRpcResult<Value> {
+        let method = match self.module_id {
+            Some(module_id) => ApiMethod::Module(module_id, method.to_string()),
+            None => ApiMethod::Core(method.to_string()),
+        };
+
+        self.connections
+            .request(peer_id, method, params.clone())
+            .await
+            .map_err(|e| JsonRpcClientError::Custom(e.to_string()))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ReconnectClientConnections {
+    connections: BTreeMap<PeerId, ClientConnection>,
+    #[allow(dead_code)]
+    tg: TaskGroup,
+}
+
+impl ReconnectClientConnections {
+    pub fn new(connector: &DynClientConnector, tg: TaskGroup) -> Self {
+        ReconnectClientConnections {
+            connections: connector
+                .peers()
+                .into_iter()
+                .map(|peer| (peer, ClientConnection::new(peer, connector.clone(), &tg)))
+                .collect(),
+            tg,
+        }
+    }
+
+    async fn request(
+        &self,
+        peer: PeerId,
+        method: ApiMethod,
+        request: ApiRequestErased,
+    ) -> anyhow::Result<Value> {
+        self.connections
+            .get(&peer)
+            .expect("Could not find client connection for peer {peer}")
+            .connection()
+            .context("Peer is disconnected")?
+            .request(method, request)
+            .await
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ClientConnection {
+    receiver: watch::Receiver<Option<DynClientConnection>>,
+}
+
+impl ClientConnection {
+    fn new(
+        peer: PeerId,
+        connector: DynClientConnector,
+        task_group: &TaskGroup,
+    ) -> ClientConnection {
+        let (sender, receiver) = watch::channel(None);
+
+        task_group.spawn_cancellable(
+            format!("peer-api-reconnection-{peer}"),
+            async move {
+                loop {
+                    let connection = retry("reconnect-peer-api", api_networking_backoff(), || {
+                        connector.connect(peer)
+                    })
+                    .await
+                    .expect("Unlimited retries");
+
+                    info!(target: LOG_CLIENT_NET_API, "Connected to peer api");
+
+                    if sender.send(Some(connection.clone())).is_err() {
+                        break;
+                    }
+
+                    connection.await_disconnection().await;
```

Comment:
Looking at it more calmly again, I think having a separate task that connects is an elegant way to express merging of shared reconnecting state.

The only different/complication in the behavior here is making reconnecting itself lazy - not attempt reconnecting, unless something signaled from `Self::connection` that it actually wants a connection. E.g. `watch` channel that the caller of `connection` sets if the underlying connection is `None`, and the loop here would for `.changed()` before doing another reconnect.